keps/sig-network/2157-kube-proxy-config/README.md

```diff
@@ -0,0 +1,155 @@
+- - Drawbacks
+- - Alternatives
+- <!-- /toc -->
+
+## Release Signoff Checklist
+
+Items marked with (R) are required *prior to targeting to a milestone / release*.
+
+- [ ] (R) Enhancement issue in release milestone, which links to KEP dir in [kubernetes/enhancements] (not the initial KEP PR)
+- [ ] - [ ] (R) KEP approvers have approved the KEP status as `implementable`
+- [ ] - [ ] (R) Design details are appropriately documented
+- [ ] - [ ] (R) Test plan is in place, giving consideration to SIG Architecture and SIG Testing input
+- [ ] - [ ] (R) Graduation criteria is in place
+- [ ] - [ ] (R) Production readiness review completed
+- [ ] - [ ] Production readiness review approved
+- [ ] - [ ] "Implementation History" section is up-to-date for milestone
+- [ ] - [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
+- [ ] - [ ] Supporting documentation—e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://git.k8s.io/enhancements
+[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
+[kubernetes/website]: https://git.k8s.io/website
+
+## Summary
+
+The current kube-proxy flags are largely ignored by default, but they **are** validated.  This is self-consistent but misleading.  As an example, you can't send `--blah`, but you can say `--proxy-mode=thisiswrong`.  The latter is essentially a no-op (it should break the kube-proxy and "fail closed" as tim might say :).
+
+We thus propose a few simple consistency amendments to the input handling of the kube-proxy which will be self-consistent and prevent "no-op" inputs that are silently ignored, and in general, suggest transparency in terms of what inputs to kube-proxy are being utilized.
+
+## Motivation
+
+The latter command ALSO ignores correct flags `--proxy-mode=iptables`, as you might expect, meaning that users might try to configure something and then spend several minutes trying to figure out why they're config wasnt exercised.
+
+Since there are several kube-proxy options that might be overriden (there are ~50 options configured in a typical config-map for the kube proxy), this is obviously a compounding problem.
+
+## Goals
+
+We thus suggest the following modifications to kube-proxy input handling.
+
+1) For some options (such as the `mode`) the configuration is printed.  We suggest that all options be printed, this is done, for example, already for the `mode` option.
+```
+I1119 16:23:58.582572       1 server_others.go:186] Using iptables Proxier.   
+```
+currently options like proxy-mode are not processed at all via command line, EVEN IF there is nothing in the configMap
+
+2) Command line options should be explicitly rejected (if ignored *or* if incorrect).
+
+3) Some subset of config options (to be iterated and added over time) can be used to **override** the input kube-proxy config map.
```

Comment:
If we're in favor of adopting the instance-specific component config described in https://github.com/kubernetes/enhancements/pull/1439, then we probably will end up rejecting all flags. My concern for adding flags (to overwrite config map values) is it's destabilizing the interface for users and could lead to confusion, also that [KEP](https://github.com/kubernetes/enhancements/blob/61d6935979ea20a235e76b0e572d470844e1dc2e/keps/sig-cluster-lifecycle/wgs/2141-instance-specific-componentconfig/README.md#why-not-just-keep-using-flags-for-instance-specific-parameters) mentioned some issues related to flags.

I think maybe we can start by rejecting the flags that implicitly got ignored as the first step, and keep an eye on https://github.com/kubernetes/enhancements/pull/1439.

Reply:
I don't think we should ignore flags, but overwrite ALL config options until the final deprecation occurs.

The final cleanup of the flags will happen only in the graduation of the KubeProxyConfiguration API. In the meantime, a warning is raised.

Reply:
The evolution of the library is to let ALL flags take precedence to configmap options until final removal.
In the meantime, a warning is raised for the user. 

I'm not sure if it's a good idea to reject existent flags explicitly,  but instead, let it overwrite CM options as a final goal of the KEP. This can be a breaking change for something that is already used wrongly, so the migration of this can be made gradually.

The instance config flags will not solve this problem since the flags precedence will still exist, in order of values precedence:
 
Flags -> KubeProxyConfigurationInstance -> KubeProxyConfiguration

Reply:
> The evolution of the library is to let ALL flags take precedence to configmap options until final removal.

Merging flags and config has proven extremely confusing in the kubelet... I would not recommend that direction. Saying that if a config file is specified it is used ~exclusively, and rejecting flags that are set along with the config file that are not honored is simpler to reason about

Reply:
@liggitt I'm probably missing some history here and mistaken the precedence of flags as an evolution of the API 

> Saying that if a config file is specified it is used ~exclusively, and rejecting flags that are set along with the config file that are not honored is simpler to reason about

This can make sense if we explicitly inform the user that flags do not have effect when the config set the field.

The main issue nowadays is the opposite of, the default values from the config overwrite the flags values (and probably this is the reason the flag precedence was introduced), so if you set a --config file like:

```
apiVersion: kubeproxy.config.k8s.io/v1alpha1
kind: KubeProxyConfiguration
```

and set a `--proxy-mode=ipvs`, the final config object ends up with `config.Mode = ""`, what's is not the expected behavior.

Reply:
kube-proxy was one component that did this wrongly. flags overriding the config is the more common pattern in k8s and outside of it. but that is still problematic and has implications around the instance specific vs global config topic.

One of the motivations for creating WG Component Standard was to help drive work in component owning SIGs to start phasing out flags in favor of config only. The WG is not very active currently and is looking for new leads.

Reply:
> kube-proxy was one component that did this wrongly. flags overriding the config is the more common pattern in k8s and outside of it. but that is still problematic and has implications around the instance-specific vs global config topic.

Is the issue only around the instance-specific vs global config? If so, what they are?

We could get a reasonable approach in the patch merge on [KEP 1439](https://github.com/kubernetes/enhancements/pull/1439) including the flag precedence for Kubelet at least.